### ISSUE TITLE
PROPOSAL: Native N-Dimensional Slices in Odin

### ISSUE THREAD

**Issue by F0x1fy** [odin-lang/Odin#542]
# Proposal: [Native N-Dimensional Slices (NDS) in Odin]

Author(s): [F0x1fy](https://github.com/F0x1fy)
Co-Author(s): [Tetralux](https://github.com/tetralux)
Last updated: [1/17/2020]

TL;DR header at bottom supplies a much shorter version. The TL;DR will give a basic overview of the proposal, and the meat of the understanding is under the [Implementation](#Implementation) header

## Abstract
The following description proposes a native inclusion of n-dimensional slices in Odin through the use of a simple and efficient algorithm and compile-time code injection instead of double-indirection.

## Background
The issue with common implementations of NDS's is double indirection. This causes accessing NDS's quite slow and inefficient. Though this implementation does have some advantages (such as jagged slices), I believe these would not benefit the Odin language well as opposed to using the more algorithmic approach.

## Proposal
This proposes the use of a specific algorithm to take n-dimensional slices and sprawl them over one dimension. This would use the (or an alteration of the) implementation from my [Sprawl](https://github.com/F0x1fy/odin-sprawl) repository as a reference.

## Rationale
An easier and simple way to use an NDS would be likely to help from detracting newer programmers, help current Odin users, and be an efficient way to deal with the issue.

## Compatibility
This would be compatible with previous versions of Odin and Odin source code as this would be an addition to the slice datatype, not an alteration. This would add the ability to access slices like so: `nds[x][y][z]` or like `nds[x, y, z]` (depending on Bill's personal preference). This would not at all change single-dimension slice access.

## Implementation
The formulae used could be easily generated at compile time. Here's the basic concept:
`n * sizes of every index smaller than n + y * sizex + x`
So then `arr[x, y, z]` would become `z * sizey * sizex + y * sizex + x` at compile-time.

Here's the runtime procedure version I created in my library [Sprawl](https://github.com/F0x1fy/odin-sprawl):
```
// Returns index from specified indexes and sizes: slice[sprawl(indexes, sizes)]
sprawl :: inline proc(indexes, sizes: [$N]$T) -> T {

    // Calculate the offset using y * sizex + x
    output := indexes[1] * sizes[0] + indexes[0];

    // Iterate every value except the last two since they're used in the previous statement
    inline for i in 0..<len(indexes) - 2 {
        mul: T = 1;

        // Multiply the sizes together, up to the count of i - 1, essentially getting the previous dimensions (z would give sizey and sizex, for example)
        for j in 0..<len(sizes) - i - 1 {
            mul *= sizes[j];
        }

        // Multiply the index by the product of the previous sizes of the dimensions
        output += indexes[len(indexes) - i - 1] * mul;
    }

    return output;
}
```

So then, using the `[x,y,z,...]` syntax, that would get translated to `sprawl([?]{x, y, z, ...}, [?]{sizex, sizey, sizez, ...})` if using my procedure as an example, however at compile time it would output as described previously.

Note that the compile-time code injection approach would require more work to make compatible with allowing dimension and boundary mutation to allow for resizing.

Further notes on more procedures can be found in [Sprawl](https://github.com/F0x1fy/odin-sprawl), though the exact formulae at the end of the README need to be updated.

## TL;DR
I propose a simple syntatic way to access n-dimensional slices (NDS's) through compile-time code injection using an algorithm from my [Sprawl](https://github.com/F0x1fy/odin-sprawl) library as a reference. Instead of the messy syntax of something like `arr[sprawl([?]{x, y, z}, [?]{sizex, sizey, sizez})]`, we'd instead use `arr[x, y, z]` or `arr[x][y][z]`.

My implementation from [Sprawl](https://github.com/F0x1fy/odin-sprawl):
```
// Returns index from specified indexes and sizes: slice[sprawl(indexes, sizes)]
sprawl :: inline proc(indexes, sizes: [$N]$T) -> T {

    // Calculate the offset using y * sizex + x
    output := indexes[1] * sizes[0] + indexes[0];

    // Iterate every value except the last two since they're used in the previous statement
    inline for i in 0..<len(indexes) - 2 {
        mul: T = 1;

        // Multiply the sizes together, up to the count of i - 1, essentially getting the previous dimensions (z would give sizey and sizex, for example)
        for j in 0..<len(sizes) - i - 1 {
            mul *= sizes[j];
        }

        // Multiply the index by the product of the previous sizes of the dimensions
        output += indexes[len(indexes) - i - 1] * mul;
    }

    return output;
}
```

This would be generated at compile time. However, keep in mind it'd take extra work to make compatible with allowing dimension and boundary mutation to allow for resizing.

**Comment by F0x1fy** [comment]
I should have stated this explicitly, my bad:
The AMOUNT of dimensions are known at compile time, and that's what's used to generate the code injection effectively.

**Comment by gingerBill** [comment]
From our discussion from the Discord, there are two "flavours" of multidimensional slices.

The most generic form would have this layout.
```
Raw_N_Dim_Slice :: struct(T: typeid, N: int) {
    data: ^T,
    lengths: [N]int,
    strides: [N-1]int,
}
```

The reason there can be multiple flavours is because a 1-d slice has an implicit stride for its length of "1" element.

**Comment by F0x1fy** [comment]
In this case, would `strides` be the offsets? Is there any issue with that struct and the 1D slices?

**Comment by F0x1fy** [comment]
Just checking in on this and seeing what your opinion is on it, and whether you're looking to potentially add it or not
